Paper: A Reproducible Data Lakehouse for High-Resolution Gastric Cancer Epidemiology Study in Chile (#1240)
scipy-conference/scipy_proceedings: papers/rene_lagos/main.md

---
# Ensure that this title is the same as the one in `myst.yml`
title: A Reproducible Data Lakehouse for High-Resolution Gastric Cancer Epidemiology Study in Chile
abstract: |

>   Cancer is the main cause of mortality in Chile since 2019. National surveillance of cancer incidence, generating high-resolution insights often requires merging massive and heterogeneous datasets. For a PhD researcher in Public Health set to understand the spatio-temporal trends of cancer incidence, the lack of high-performance server infrastructure and the limitations of traditional epidemiological tools when handling longitudinal Big Data on local machines can be important barriers.

Reviewer: This abstract is a strong start, but I think a little reorganizing and prioritizing active voice would improve it further. 
> Since 2019, cancer has been the leading cause of mortality in Chile. Leveraging national surveillance of cancer incidence to understand the factors behind this trend typically requires merging massive heterogeneous datasets. Unfortunately, the lack of high-performance server infrastructure and the limitations of traditional epidemiological tools for handling big data on local machines poses a significant challenge.

Author: Thanks @samuelcwilliams , I included your suggestion

>   We developed a portable, open-source Data Lakehouse architecture built entirely on the Python ecosystem to study Gastric Cancer incidence in the Chilean population. We obtained 21 years of longitudinal, anonymized data of Chilean population (20 million inhabitants), mortality records, hospital discharges, and insurance claims by age, sex, and county, which we stored using Parquet files and Google Drive provided by the University of Chile. Inspired on a Dynamic-ETL approach (Ong et al. 2017), we implemented a modular ELT workflow using Jupyter Notebooks and Google Colab, where each notebook performed a specific extraction, normalization or loading process for each dataset. We employed rpy2 and mgvc packages for age-standardized time-series modeling, GeoPandas and Sci-kit for spatial analysis and identification of high-risk zones; and NetworkX to model patient-consultation trajectories, identifying gastroenterology referral clusters in the national health system. Finally, we developed a Streamlit dashboard that allows team members and the general public to visualize cancer incidence time-series charts and referrals clusters dynamically.

Reviewer: This section is fantastic. One small fix, I believe "Inspired on a Dynamic-ETL approach" was meant to be "Inspired by a Dynamic-ETL approach".

Reviewer 2: Not sure if this is intentional, but the styling of package names is different here than later in the paper (e.g. rpy2 vs `rpy2`)

Author: Fixed "Inspired by a Dynamic-ETL approach" and `rpy2`

  This architecture supported the submission of three conference abstracts and a manuscript to a public health journal, ensuring fully reproducible and documented results.  This project demonstrates that high-resolution, national-level public health research does not require expensive enterprise software. By utilizing a version-controlled, cloud-collaborative, and nearly zero-cost Python stack, we provide a blueprint for researchers in resource-limited settings to conduct reproducible epidemiological surveillance.

---

## Introduction

> Cancer is a priority health problem in Chile. It is currently the leading cause of death and the main burden of disease in the country [@doi:10.4067/S0034-98872021000100149]. Among cancers, gastric cancer (GC) is particularly challenging, as Chile has one of the highest incidence and mortality rates in the world, along with some East Asian countries [@ferlay_j_global_2024].

Reviewer: This introduction might flow a little better with something like:
> Cancer is a high-priority health problem in Chile .It is currently the leading cause of death and the main burden of disease in the country. Gastric cancer (GC) is particularly prevalent, with Chile having one of the highest incidence and mortality rates in the world alongside some East Asian countries.

Author: Included suggestion 👍

Chile maintains population-based cancer registries (PCRs) that collect information on all cancer cases across a set of regions for epidemiological surveillance and planning cancer control policies. These registries cover approximately 20% of the national population; however, estimates are reported at the national level, and updated figures have not been available since 2019.

> We set out to estimate updated estimates of GC incidence at the provincial level to support the planning and evaluation of prevention policies. This required applying state-of-the-art methodologies, which demand a large amount of longitudinal data on mortality, hospital discharges, population by sex and age, and socioeconomic variables.

Reviewer: "We set out to update estimates of GC incidence" might flow a little better.

Author: Included suggestion 👍

> Handling this data with programming languages typically used in epidemiology, such as Stata and R, posed a significant computational risk. Public health researchers often conduct studies using the same hardware used for office tasks, as they use statistics aggregated at the national level. For higher geographic granularity and longer period studies, as in this case, high-performance computing was necessary but not available. Furthermore, tools like R and RStudio use in-memory processing, meaning the entire dataset must be loaded into the computer's RAM for analysis. When handling two decades of longitudinal data for a population of 20 million inhabitants—including heterogeneous records such as mortality, hospital discharges, and insurance claims—the resulting datasets can easily exceed dozens of gigabytes. Attempting to merge and normalize these massive datasets on a typical laptop or office computer could have led to system crashes or "out of memory" errors.

Reviewer: Are these data operations purely memory-bound, or is there also a speed/performance component as well? If so, that might be worth highlighting here alongside the memory issues.

Author: I was only a memory issue in this case

To manage the complexity of transforming heterogeneous health records, we adopted the Dynamic ETL (D-ETL) framework [@ong_dynamic-etl_2017]. Traditional GUI-based ETL tools often lack the flexibility required for the complex requirements of health data and suffer from a lack of transparency in their underlying transformation logic. Following the D-ETL approach, we modularized our data processing by mapping and transforming each dataset (e.g., mortality records, hospital discharges) separately. This approach allowed us to automate repetitive tasks using scalable, reusable, and customizable Python code while retaining manual control over complex mappings for epidemiological analysis.

### Workflow Implementation

> We took advantage of University of Chile´s Google Workspace to use low-cost cloud storage in Google Drive. This also ensured that data storage complied Universities privacy and security policies. We created three main folders: 1) Original Data, for the original files (zip, xls, csvs, etc) and documentation received and organized in subfolders by data source, 2) Raw Data, for the data extracted into optimized format for intake, and 3) Data Warehouse, for normalized and standardized data. Additionally, we created a separate folder for each analysis, containing the data transformation script and the statistical analysis.

Reviewer 2: ```suggestion
We took advantage of University of Chile´s Google Workspace to use low-cost cloud storage in Google Drive. This also ensured that data storage complied with the University's privacy and security policies. We created three main folders: 1) Original Data, for the original files (zip, xls, csvs, etc) and documentation received and organized in subfolders by data source, 2) Raw Data, for the data extracted into optimized format for intake, and 3) Data Warehouse, for normalized and standardized data. Additionally, we created a separate folder for each analysis, containing the data transformation script and the statistical analysis.
```

:::{figure} figure1.png
:label: fig:stream
Tools used to implement Dynamic Extraction-Load-Transform workflow.
:::

### Data model and sources

We used an Effective Coverage framework that distinguishes populations, needs, resources, production and health outcomes [@marsh_effective_2020] and the Chilean Norm for Health Information for data modeling [@departamento_de_estadistica_e_informacion_de_salud_norma_2016; @departamento_de_estadisticas_e_informacion_de_salud_norma_2023]. Figure 2 shows the entities, attributes and relationships used to model GC prevention that were stored in the data lake. Figure 3 shows the star-schema for the data cube (multidimensional table) used to analyze GC incidence.

:::{figure} figure2.png
:label: fig:em
Entity-Relationship Diagram.
Colors: blue=populations; yellow=resources; brown=production; green=needs; red=health outcomes.
*Individual-level data.
**Linkable individual-level data.
:::

:::{figure} figure3.png

> :label: fig:em

Reviewer 2: The figure label here is dublicated so it shows up as Figure 2

Author: Fixed figures labels

Star schema of data cube used to analyze gastric cancer incidence in Chile 2003-2024.
:::

We retrieved GC cases (ICD-10 category "C16") from the PBCR database (1998–2019) via the Ministry of Health website. Upon request, the Department of Health Statistics and Information (DEIS) provided individual-level records for GC deaths and hospital discharges (2003–2024). We obtained age- and sex-specific population projections and rurality data from the National Statistics Institute (INE). Finally, we sourced insurance coverage data from public health insurer FONASA (2003–2024), and from the Superintendency of Health for ISAPRE beneficiaries (2010–2024); poverty percentages from the Ministry of Social Development; and consultations and UGE production in public hospitals from monthly statistical reports from DEIS (2009-2024).

### Statistical methods and packages

> To estimate GC incidence, we used mortality-incidence ratios [@chatignoux_how_2021] implemented with Quasi-Poisson regression models with R's `mgcv` library [@wood_mgcv_2000]. We embedded R scripts in Python code using the `rpy2` library [@gautier_rpy2_2023]. We used `NetworkX` to generate health center networks for GC diagnosis, based on referral patterns [@hagberg_exploring_2008]. We used `Geopandas` for maps and `Streamlit` to publish a dashboard [@khorasani_web_2022,].

Reviewer: This sounds like a really fascinating use of `NetworkX`. I think including a figure resulting from that would help highlight this even more.

Reviewer 2: Is there a reference for Geopandas?

Author: Included Geopandas reference and referral network figure 👍

## Results

> Table 1 shows the Jupyter notebooks developed. For extraction, since each original dataset had a unique script associated, notebooks were stored in the Original Data folder, in their corresponding subfolder (figure 1). During the loading stage, each normalized entity had its own script, so notebooks were stored in the Data Warehouse folder.

Reviewer: The second sentence here might flow a little better without "For extraction".

Author: Included suggestion

```{list-table} Jupyter notebooks developed.
:label: tbl:materials
:header-rows: 1
* - Original Data *Extraction*
  - *Loading* to Data Warehouse
  - *Transformation* Multidimensional Tables
* - Population estimates, Poverty estimates, FONASA Beneficiaries, ISAPRE Beneficiaries, Health center population, Health centers, Production, Claims, Referrals, Population-Based Cancer Registry, Hospital discharges, Deaths, Administrative maps

>   - District Population, FONASA Beneficiaries, ISAPRE Beneficiaries, Health Center Population, Health centers, Consultation production, Endoscopy production, Cancer referals, Non-cancer referrals, Gastric cancer cases, Hospital discharges, Deaths, DIM Age, DIM Sex, DIM Districts, DIM Cancer, GEO Provinces, GEO Health services.

Reviewer 2: ```suggestion
  - District Population, FONASA Beneficiaries, ISAPRE Beneficiaries, Health Center Population, Health centers, Consultation production, Endoscopy production, Cancer referrals, Non-cancer referrals, Gastric cancer cases, Hospital discharges, Deaths, DIM Age, DIM Sex, DIM Districts, DIM Cancer, GEO Provinces, GEO Health services.
```

      <td>Codigo, Nombre, Nivel, Tipo, Complejidad, Comuna, Servicio, Latitud, Longitud, Nombre Comuna, Nombre Region, Porc_rural_comuna, Destino GES</td>
      <td>4582</td>
      <td>465.5 KB</td>
      <td>218 KB</td>
    </tr>
    <tr>
      <td></td>
      <td>CUBO_DERIVACIONES_MDD</td>
      <td>Origen, Destino, Código prestación, COMGES, GES, Derivaciones</td>
      <td>6762</td>
      <td>317.1 KB</td>
      <td>33 KB</td>
    </tr>
    <tr>
      <td></td>
      <td>CUBO_INSCRITOS_DERIVACIONES</td>
      <td>Origen, Sexo, Rango Edad, Año, Inscritos, Código prestación, Prestación, Tipo prestación, Especialidad, GES, Derivaciones</td>
      <td>906782</td>
      <td>76.1 MB</td>
      <td>576 KB</td>
    </tr>
    <tr>
      <td></td>
      <td>CUBO_COBERTURA_SERVICIOS</td>
      <td>idServicio, Codigo, RangoEdad4080, Año, Casos, Poblacion, PoblacionRural, IncidenciaEst, Consultas, Endoscopias, Confirmacion_C16, Confirmacion_HP, Tratamiento_C16, Tratamiento_HP, Egresos, Defunciones, Isapre, Fonasa A, Fonasa B, Fonasa C, Fonasa D, Servicio, Macrorregion, Sexo, Periodo</td>
      <td>10424</td>
      <td>2.4 MB</td>
      <td>429 KB</td>
    </tr>
  </tbody>
</table>
:::

> Each analysis folder was replicated multiple times. The GC incidence scripts were replicated to prepare a congress poster and a scientific journal manuscript [@lagos_como_2025]. This last version was published in a [Github repository](https://github.com/rlagosb/GastricCancerIncidence) as part of a manuscript publication. The diagnosis network scripts were replicated three times: first to generate a report for the Ministry of Health, second to present a poster at a congress, and third to prepare a scientific journal manuscript [@lagos_caracterizacion_2025]. And the diagnosis coverage scripts were replicated for a congress poster and a manuscript [@lagos_effective_2026]. The dashboard was published in streamlit cloud and used in congress presentations, but most oftenly for personal use: [tesisrenelagos.streamlit.app](https://tesisrenelagos.streamlit.app/).

Reviewer: This section might benefit from being broken up into individual sections for the GC incidence scripts, diagnosis network scripts, and diagnosis coverage scripts.

Reviewer 2: I was a bit confused why the emphasis on the analysis being replicated. It may be helpful to start with a sentence clarifying what this shows (data warehouse has already been useful for many analyses?), before going into the specifics of the scripts.

To manage the complexity of heterogeneous records, we adopted a D-ETL approach, implementing a "one dataset, one notebook" model. This provided the transparency of Python often lacking in GUI-based tools, where underlying transformation logic is frequently hidden. By embedding rules, assumptions, and mapping specifications as text cells directly alongside executable code, we created a single, human-readable file for each target dataset. This integrated format facilitated an iterative process of validation and debugging, ensuring that every step of the analytical logic remained visible and easily scrutinized.

> By using Google Colab and Drive, we established a portable, cloud-native infostructure that bypassed the lack of high-performance server capability often found in public health research. This ready-to-use environment turned standard cloud tools into a secure environment, ensuring that the analysis of data complied with institutional security policies while providing a shared, collaborative platform for researchers. To navigate the 12 GB RAM and 50 GB storage limits of the free-tier environment, the partitioning of the pipeline into 38 discrete Jupyter Notebooks  avoided a memory-intensive scripts to an iterative, lightweight process that effectively managed 21 years of records for 20 million inhabitants.

Reviewer: The last sentence was a little hard for me to understand. It may be worth reformatting to something like:

"By partitioning the pipeline into 38 discrete Jupyter Notebooks we avoided hitting the 12GB RAM and 50GB storage limits of the free-tier environment while providing an iterative, lightweight process that effectively managed 21 years of records for 20 million inhabitants.

Author: Included suggestion

This cloud-native environment facilitated statistical workflow through the `rpy2` library; we leveraged Python’s high-performance ecosystem for data engineering, while transitioning seamlessly into R for specialized Quasi-Poisson incidence modeling. This interoperability eliminated the need for manual, error-prone data exports between software environments.

A critical choice in this implementation was the use of the Parquet columnar storage format, which provided substantial storage efficiency and performance gains. For example, a population dataset with nearly two million entries was compressed to just 3.2 MB on disk. Beyond speed, Parquet’s ability to store text metadata solved persistent encoding issues with Spanish characters in Chilean records, thereby safeguarding data integrity across heterogeneous sources.

> Because the system follows the Effective Coverage framework, it can be easily replicated for colorectal or prostate cancer by simply mapping the "Need" and "Production" required datasets. Relational integrity is maintained through the data model based on the Chilean Norm for Health Information, ensuring that standardized attributes—such as sex, age, and county IDs—remain consistent across heterogeneous sources. This design allows for expansion by adding new records to existing tables with minimal structural changes. Furthermore, the strategic implementation of common Data Warehouse dimensions, such as DIM_Age, enables the integration of new datasets with varying granularities without requiring a rewrite of the core transformation logic.

Reviewer: Are there any plans to replicate or expand this study to investigate other forms of cancer? If so, I think the paper would benefit from including a "Future Work" section that discusses this in more detail.

Author: Not clear at the moment, thank you for the suggestion!

### Limitations

> While the free version of Google Colab facilitates access, it necessitates a rigorous pre-loading anonymization process to ensure no Protected Health Information enters the public cloud environment. a standard implementation for non-anonymized records would require an ETLT (Extract-Transform-Load-Transform) workflow that enables a pre-load transformation stage to mask or encrypt sensitive identifiers before data reaches a cloud-based staging area, ensuring compliance with privacy regulations.

Reviewer 2: ```suggestion
While the free version of Google Colab facilitates access, it necessitates a rigorous pre-loading anonymization process to ensure no Protected Health Information enters the public cloud environment. A standard implementation for non-anonymized records would require an ETLT (Extract-Transform-Load-Transform) workflow that enables a pre-load transformation stage to mask or encrypt sensitive identifiers before data reaches a cloud-based staging area, ensuring compliance with privacy regulations.
```
